lib/ExecutionEngine/ExecutionEngine2.cpp

```diff
@@ -0,0 +1,216 @@
+  setBackend(backend);
+}
+
+/// Set the code generator kind to \p backend.
+void ExecutionEngine::setBackend(llvm::StringRef backend) {
+  setBackend(createBackend(backend));
+}
+
+/// Set the code generator to the given \p backend.
+void ExecutionEngine::setBackend(Backend *backend, bool ownsBackend) {
+  module_.reset(new Module);
+  rawModule_ = module_.get();
+  if (ownsBackend_) {
+    delete backend_;
+  }
+  backend_ = backend;
+  ownsBackend_ = ownsBackend;
+  clear();
+
+  if (hostManager_) {
+    EXIT_ON_ERR(hostManager_->clearHost());
+    hostManager_.reset();
+  }
+
+  if (backend) {
+    std::vector<std::unique_ptr<runtime::DeviceConfig>> configs;
+    auto config =
+        llvm::make_unique<runtime::DeviceConfig>(backend->getBackendName());
+    configs.push_back(std::move(config));
+    hostManager_ = llvm::make_unique<runtime::HostManager>(std::move(configs));
+  }
+}
+
+const Backend *ExecutionEngine::getBackend() const { return backend_; }
+
+ExecutionEngine::~ExecutionEngine() {
+  // Call setBackend to make sure that backend_ is deleted if it's owned.
+  setBackend(nullptr, /*ownsBackend*/ false);
+}
+
+void ExecutionEngine::clear() {
+  if (hostManager_) {
+    EXIT_ON_ERR(hostManager_->clearHost());
+  }
+  compiledFunctions_.clear();
+}
+
+void glow::updateInputPlaceholders(PlaceholderBindings &bindings,
+                                   llvm::ArrayRef<Placeholder *> ph,
+                                   llvm::ArrayRef<Tensor *> inputs) {
+  assert(inputs.size() == ph.size() &&
+         "The number of inputs does not match the number of Placeholders");
+
+  for (int i = 0, e = ph.size(); i < e; i++) {
+    assert(ph[i] && "Invalid value");
+    auto *backingTensor = bindings.get(ph[i]);
+    assert(backingTensor && "Can't find the placeholder");
+    auto dim = inputs[i]->dims();
+    (void)dim;
+    assert(backingTensor->getType().isEqual(inputs[i]->getType()) &&
+           "Mismatch on Placeholder and Tensor types.");
+    backingTensor->assign(inputs[i]);
+  }
+}
+
+void glow::updateInputPlaceholdersByName(PlaceholderBindings &bindings,
+                                         Module *mod,
+                                         llvm::ArrayRef<llvm::StringRef> ph,
+                                         llvm::ArrayRef<Tensor *> inputs) {
+  assert(inputs.size() == ph.size() &&
+         "The number of inputs does not match the number of Placeholders");
+
+  for (int i = 0, e = ph.size(); i < e; i++) {
+    Placeholder *p = mod->getPlaceholderByName(legalizeName(ph[i]));
+    Tensor *t = inputs[i];
+    assert(t && "Invalid tensor.");
+    assert(p && "Invalid placeholder.");
+    updateInputPlaceholders(bindings, {p}, {t});
+  }
+}
+
+void ExecutionEngine::runInternal(ExecutionContext &context,
+                                  llvm::StringRef name) {
+  // Make sure that the bindings have backing tensors for all placeholders.
+  context.getPlaceholderBindings()->allocate(rawModule_->getPlaceholders());
```

Comment:
Should this be a call on HostManager rather than requiring we store a raw pointer to what we passed in? Alternatively, is it better to keep a copy of the full bindings, clone it and merge it?

Reply:
I think on the HostManager is a bit general, what do you think about just removing this call entirely and letting the caller ensure their bindings are properly allocated?